### ISSUE TITLE
YDoc sync via Y.applyUpdate does not work on special cases

### ISSUE THREAD

**Issue by Manuelbaun** [yjs/yjs#175]
Hi, 
first of all, thank you for developing _yjs_. I think it is a great library and actually very easy to use. 

I followed your examples on applying the update on docs, from a local _ydoc_ instance to a remote instance and vice versa. It worked fine, but then I tried to organize my code more into TS classes and suddenly I encountered an issue.

consider the following code, I tired to make the example as small as possible:
```ts
/**
 * Remote doc updated form local doc does not work, 
 * when map is set in the class, for example:
 * this.map.set("time", 60);
 */

function runPlayground() {
	let local: Test;
	let remote: Test;

	// Setup the test classes
	local = new Test({
		yDoc: new Y.Doc(),
		name: 'local',
		emitChange: (data) => {
			remote.onReceive(data);
		}
	});

	remote = new Test({
		yDoc: new Y.Doc(),
		name: 'remote',
		emitChange: (data) => {
			local.onReceive(data);
		}
	});

	// Add data to be synced
	const msg = {
		user: 'Bro',
		time: Date.now(),
		message: 'Hallo'
	};
	console.log('Add Object to local doc', msg);
	local.add(msg);
}

class Test {
	name: string;
	arr;
	map;
	yDoc;
	constructor({ yDoc, name, emitChange }) {
		this.yDoc = yDoc;
		this.name = name;
		this.arr = yDoc.getArray('arr');
		this.map = yDoc.getMap('map');

		try {
			// <== Uncomment this and Y.applyUpdate(remote, update)
                        // does not apply the update to the remote yDoc instance

			// this.map.set('time', 60);
		} catch (err) {
			console.log(err);
		}

		this.yDoc.on('update', (update) => {
			console.log(`==> ${this.name} Array now:`, this.arr.toArray());
			console.log(`==> ${this.name} Update emit`, update);
			emitChange(update);
		});
	}

	add(msg: Object) {
		this.arr.push([ msg ]);
	}

	onReceive(update) {
		console.log(this.name, 'received update', update);
		Y.applyUpdate(this.yDoc, update);
	}
}
```

So what've did:
I created a Test class and passed on the ydoc instance.  Then, with ydoc I create, an YArray and a YMap. The class uses then the Y API to change the YMap and the YArray, like append to the array or set the fields of the map.

Next, when I try to set the map via map.set('time',60), then the strange behavior happens:
1. local.add(msg)
2. local.on => gets triggered and 'sends' to the remote yDoc
3. remote yDoc receives via `remote.onReceive`
4. Then Y.applyUpdate(this.yDoc, update)  => should apply the update to the remote yDoc
5. !!!!! here the remote.on('update', ...) should trigger, but that does not happen and no changes are applied.

Info: I also tried to create the YArray and the YMap before and passed it on the Test class => the same result.  If I modify the map, then Y.applyUpdate(...) does not work

----------------------------------------------------
On the other hand, if I implement it like this, then it workes:
```ts
function play() {
	// create yDocs
	const local = new Y.Doc();
	const remote = new Y.Doc();

	// setup on Update listener
	local.on('update', (update) => {
		console.log('local update => apply to remote');
		Y.applyUpdate(remote, update);
	});

	remote.on('update', (update) => {
		console.log('remote update from local', remote);
		const remoteMap = remote.getMap('map');
		console.log(remoteMap.get('key1'));
	});

	// creates array and maps
	const remoteArr = remote.getArray('arr');
	const remoteMap = remote.getMap('map');

	const localArr = local.getArray('arr');
	const localMap = local.getMap('map');

	// apply changes to the maps and array
	// remoteMap.set('key1', 1234);

	localArr.push([ '12345' ]);
	localMap.set('key1', 123);
}
```

What could it be? Do I use the yDoc wrong?

Thank you very much

**Comment by Manuelbaun** [comment]
alright, I found something, not sure why it happens, but I found out, when it happens and when not. Consider the changes, in the Test constructor:

```ts
function runPlayground() {
	let local: Test;
	let remote: Test;

	// Setup the test classes
	local = new Test({
		yDoc: new Y.Doc(),
		name: 'local',
		emitChange: (data) => {
			remote.applyUpdate(data);
		}
	});

	remote = new Test({
		yDoc: new Y.Doc(),
		name: 'remote',
		emitChange: (data) => {
			local.applyUpdate(data);
		}
	});

	// Add data to be synced
	const msg = {
		user: 'Bro',
		message: 'Hallo'
	};
	console.log('Add Object to local doc', msg);
	local.add(msg);
	local.set('time', 123);

	local.setObject({
		gameID: 123,
		name: 'jo',
		ts: Date.now()
	});
	// local.add(msg);
	// local.add(msg);
}

class Test {
	name: string;
	yDoc;
	map;
	arr;

	constructor({ yDoc, name, emitChange }) {
		this.name = name;
		this.yDoc = yDoc;
		this.map = this.yDoc.getMap('map');
		this.arr = this.yDoc.getArray('arr');

                // ================INFOS =======
                // If uncomment, it still breaks and does not sync, 
                // but as long there are not map.set inside the constructor
                // the changes are also applied to the remote doc
           
		// this.map.set("breaksTheUpdate", true);


		this.yDoc.on('update', (update, origin) => {
			console.log(`${this.name} Map:`, this.map.entries());
			for (const entry of this.map.entries()) {
				console.log(entry);
			}
			console.log(`${this.name} Array:`, this.yDoc.getArray('arr').length);
			console.log(`==> ${this.name} emit Update`);
			emitChange(update);
		});
	}

	setObject(game) {
		this.yDoc.transact(() => {
			for (const key in game) {
				console.log(key, game[key]);
				this.map.set(key, game[key]);
			}
		});
	}

	set(key: string, value: any) {
		console.log(this.name, 'Set map with', key, value);
		this.map.set(key, value);
	}

	add(msg: Object) {
		this.yDoc.getArray('arr').push([ msg ]);
	}

	applyUpdate(update) {
		console.log(this.name, 'apply Update');
		Y.applyUpdate(this.yDoc, update);
	}
}
```

**Comment by dmonad** [comment]
Hi @Manuelbaun happy you like the work :-) 

Here is a bit of background of how Yjs organizes data internally: Each change on the document is represented by a unique id that somehow looks like this: `{ client: number, clock: number }`, where `client` is your unique clientid and `clock` is an increasing number. 

**change 1:** `{ client: 393939, clock: 0 }`
**change 2:** `{ client: 393939, clock: 1 }`
**change 3:** `{ client: 393939, clock: 2 }`
...

In order to apply **change 3** you need to apply all changes that happened before. Otherwise Yjs stores it and waits for the other changes to come-in. 

You implemented the update propagation part correctly. But here is the problem

```js
// change 1 is created
this.map.set("breaksTheUpdate", true);

// From now on, all document updates are propagated.
// But they can't be applied because change1 is still missing
this.yDoc.on('update', (update, origin) => {
	console.log(`${this.name} Map:`, this.map.entries());
	for (const entry of this.map.entries()) {
		console.log(entry);
	}
	console.log(`${this.name} Array:`, this.yDoc.getArray('arr').length);
	console.log(`==> ${this.name} emit Update`);
	emitChange(update);
});
```

Your example will likely work when you change the order:

```js
this.yDoc.on('update', (update, origin) => {
	console.log(`${this.name} Map:`, this.map.entries());
	for (const entry of this.map.entries()) {
		console.log(entry);
	}
	console.log(`${this.name} Array:`, this.yDoc.getArray('arr').length);
	console.log(`==> ${this.name} emit Update`);
	emitChange(update);
});

this.map.set("breaksTheUpdate", true);
```

But it would be even better when you synced the documents after establishing the connection (I guess you plan to propagate the document updates via some communication layer). You can do so as described in this section https://github.com/yjs/yjs#example-sync-two-clients-by-exchanging-the-complete-document-structure

**Comment by Manuelbaun** [comment]
Hi @dmonad, 
thank you for your fast response and for clarifying the misconceptions I had in regards to the changes. It is exactly like you said. So what I ended up doing is: when I establish a connection, I send the whole document state and then there are no issues with syncing anymore. 

I am using WebRTC to build a small p2p game :). It's a Uni project and I like to get familiar, with the different technologies work. I found that yjs gives me no issues, except this the above mentioned, while being very fast.
